### ISSUE TITLE
Final Project Test Case 2

### ISSUE THREAD

**Issue by youc1000** [nus-cs2030/2324-s2#432]
Anyone has ever gotten this before? Idk what went wrong

![Screenshot 2024-04-03 145521](https://github.com/nus-cs2030/2324-s2/assets/164208581/50c37a6b-1a5d-4624-b82d-15590d473a94)
![image](https://github.com/nus-cs2030/2324-s2/assets/164208581/43544982-5d77-4c9d-b80b-6d85c94c94d2)

**Comment by Jeremythong2002** [comment]
The error seems to be that the queue of 1 is not being updated. If your arrive class loops through all the servers to check their availability and then loop through to check their queues availability, then this issue would indicate that the first servers' queue is erroneously being recorded as free when it should not be.

**Comment by iceflakes12** [comment]
Is this test case 3? If so then likely due to your queue being updated in the done event, but you only want your queue to be updated after the server has finished resting

happy coding
((ヽ(๑╹◡╹๑)ﾉ))♬

**Comment by tianyiiiiiii** [comment]
to add on to Jeremy's point, you can check to see where you updated the servers' queues and if you are calling the most updated queues each time you update them

**Comment by ZhengJie2001** [comment]
check whether u remove the customer from the queue properly, you may have accidentally removed the customer twice

**Comment by urmothergei** [comment]
had the same issue. might be cuz ur using the wrong serverList somewhere. just make sure your serverList you're using to add/remove customer is the most updated one. Meaning when you make new Server() always use/ take reference from the serverList that is passed into nextEvent(ImList<Server> serverList).

**Comment by Monoji77** [comment]
If i'm not wrong, I obtained the exact same error as you. My initial implementation was to have a Pair<Integer, Integer>(currentOccupancy, maximumOccupancy), where current Occupancy is initialised to 0 and maximumOccupancy is intialised to queue size + 1 (to account for customer being served). 

If your current implementation is the same/similar as described above, your code will not work simply because in your Done event, you'll probably implement the update of servers such that you'll take currentOccupancy minus 1. This will then cause your waiting servers to think that there is a server available to queue since someone "left" and there is no indication of whether a server is serving or not. 

Eg. 2 human servers, queue size 1, has rest time

current moment:
Server 1: [cust 1, cust 3]
Server 2: [cust 2, (no queue) ]

next moment when cust 1 is done serving but server 1 rests:
**CORRECT IMPLEMENTATION**
Server 1: [(not serving), cust3 ]
Server 2: [cust 2, cust 4]

**WRONG IMPLEMENTATION** (my implementation which breaks effect free programming concept)
Server 1: [(not serving), cust3, cust4]  * this works since customer 1 left and i have a "pseudo" space for cust4 based on my occupancy logic
Server 2: [cust2, (no queue)] 

TLDR, the test case fails because your code pushes a customer into the server queue despite the server resting and sitll having a full queue. The code implemented is likely to be checking that occupancy is not full: this is the case because somewhere in your Done event just minuses 1 away from occupancy without caring for the resting event.

What i did: gatta recode the implementation for your servers buddy ahha. I don't think we can get away with the lack of an indicator variable to flag that the server is serving somebody. This is especially so because of the need for human servers to rest

Hope this helps :)

**Comment by youc1000** [comment]
thanks everyone its resolved!
